src/app/reporting/ReportingEngine.cpp

```diff
@@ -0,0 +1,321 @@
+                ExitNow();
+            }
+        }
+    }
+
+    eventList.EndOfEventList();
+    SuccessOrExit(eventList.GetError() == CHIP_NO_ERROR);
+
+    event_count = CountEvents(apReadHandler, initialEvents);
+    ChipLogDetail(DataManagement, "Fetched %d events", event_count);
+exit:
+
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(DataManagement, "Error retrieving events, err = %d", err);
+    }
+
+    return err;
+}
+
+CHIP_ERROR ReportingEngine::BuildAndSendSingleReportData(ReadHandler * apReadHandler)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    chip::System::PacketBufferTLVWriter reportDataWriter;
+    ReportData::Builder reportDataBuilder;
+    chip::System::PacketBufferHandle bufHandle = System::PacketBufferHandle::New(chip::app::kMaxSecureSduLength);
+
+    VerifyOrExit(!bufHandle.IsNull(), err = CHIP_ERROR_NO_MEMORY);
+
+    reportDataWriter.Init(std::move(bufHandle));
+
+    // Create a report data.
+    err = reportDataBuilder.Init(&reportDataWriter);
+    SuccessOrExit(err);
+
+    // Fill in the EventList.
+    err = BuildSingleReportDataEventList(reportDataBuilder, apReadHandler);
+    SuccessOrExit(err);
+
+    // TODO: Add mechanism to set mSuppressResponse to handle status reports for multiple reports
+    if (apReadHandler->mSuppressResponse)
+    {
+        ChipLogDetail(DataManagement, "<RE> suppress response..");
+        reportDataBuilder.SuppressResponse(apReadHandler->mSuppressResponse);
+    }
+
+    if (mMoreChunkedMessages)
+    {
+        reportDataBuilder.MoreChunkedMessages(mMoreChunkedMessages);
+    }
+
+    reportDataBuilder.EndOfReportData();
+    SuccessOrExit(reportDataBuilder.GetError());
+
+    err = reportDataWriter.Finalize(&bufHandle);
+    SuccessOrExit(err);
+
+    VerifyOrExit(bufHandle->EnsureReservedSize(System::PacketBuffer::kDefaultHeaderReserve), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+
+#if CHIP_CONFIG_IM_ENABLE_SCHEMA_CHECK
+    {
+        ChipLogDetail(DataManagement, "<RE> Dumping report data...");
+        chip::System::PacketBufferTLVReader reader;
+        ReportData::Parser report;
+
+        reader.Init(bufHandle.Retain());
+        reader.Next();
+
+        err = report.Init(reader);
+        SuccessOrExit(err);
+
+        err = report.CheckSchemaValidity();
+        SuccessOrExit(err);
+    }
+#endif // CHIP_CONFIG_IM_ENABLE_SCHEMA_CHECK
+
+    ChipLogDetail(DataManagement, "<RE> Sending report...");
+    err = SendReport(apReadHandler, std::move(bufHandle));
+    VerifyOrExit(err == CHIP_NO_ERROR, ChipLogError(DataManagement, "<RE> Error sending out report data!"));
+
+exit:
+    ChipLogFunctError(err);
+    if (err != CHIP_NO_ERROR)
+    {
+        apReadHandler->Shutdown();
+    }
+    return err;
+}
+
+void ReportingEngine::Run(System::Layer * aSystemLayer, void * apAppState, System::Error)
+{
+    ReportingEngine * const pEngine = reinterpret_cast<ReportingEngine *>(apAppState);
+    pEngine->Run();
+}
+
+void ReportingEngine::ScheduleRun()
```

Comment:
@bzbarsky-apple ScheduleRun is called during processing read request in read handler